reference/3.0/Microsoft.PowerShell.Core/About/about_Ref.md

```diff
@@ -5,72 +5,135 @@ locale:  en-us
-the object that was passed to it. If the object is changed within the method,
-the change appears as a change in the value of the variable when control
-returns to the calling method.
+In the following example, a method attempts to change the value of a
+variable that is passed into it, and fails.
+
+```powershell
+Function Test($data)
+{
+    $data = 3
+}
+
+$var = 10
+Test -data $var
+$var
+```
 
-To use referencing, the parameter must be a reference variable. If it is not,
-an InvalidArgument exception is thrown.
+```output
+10
+```
 
-The parameters used in method invocations must match the type required by the
-methods.
+When you pass a variable *by value*, you are passing a copy of the data.
 
-Examples:
+In the following example, a variable containing a `Hashtable` is passed to a
+method. The method adds its own key, which you can still view after the
+method executes.
 
 ```powershell
-PS> function swap([ref]$a,[ref]$b)
->> {
->>     $a.value,$b.value = $b.value,$a.value
->> }
-
-PS> $a = 1
-PS> $b = 10
-PS> $a,$b
-1
-10
-PS> swap ([ref]$a) ([ref]$b)
-PS> $a,$b
-10
-1
-
-PS C:\ps-test> function double
->> {
->>     param ([ref]$x) $x.value = $x.value * 2
->> }
-
-PS> $number = 8
-PS> $number
-8
-PS C> double ([ref]$number)
-PS C> $number
-16
+Function Test($data)
+{
+    $data.Value = "New Text"
+}
+
+$var = @{}
+Test -data $var
+$var.Value
+```
+
+```output
+New Text
+```
+
+When passing a variable *by reference*, the method can change the data and
+that change will persist after the method executes.
+
+You can code your methods to take references, regardless of the type of data
+passed. When doing so, you will need to access the data passed to your
+method using the `Value` property of the
+`System.Management.Automation.PSReference` type.
+
+When calling your method, you must type cast you variable as a reference.
+The brackets and parenthesis are BOTH required.
+
+```powershell
+Function Test([ref]$data)
+{
+    $data.Value = 3
+}
+
+$var = 10
+Test -data ([ref]$var)
+$var
+```
+
+```output
+3
+```
+
+Certain .NET methods may require you to pass a variable as a reference. When
+the method's definition uses the keywords `in`, `out`, or `ref` on a
+parameter, it expects a reference.
+
+```powershell
+[int] | Get-Member -Static -Name TryParse
 ```
 
-The variable must be a reference variable.
+```output
+Name     MemberType Definition
+----     ---------- ----------
+TryParse Method     static bool TryParse(string s, [ref] int result)
+```
+
+For the `TryParse` method, you can attempt to parse a string as an integer.
+If the method succeeds, the method will return true, and the result will
+be stored in the variable you passed **by reference**.
 
 ```powershell
-PS C:\ps-test> double $number
-double : Reference type is expected in argument.
-At line:1 char:7
-+ double  <<<< $number
+$number = 0
+[int]::TryParse("15", ([ref]$number))
+$number
 ```
 
-## SEE ALSO
+```output
```

Comment:
Do not separate the output like this. Make this an illustrative example (see out style guide). It is confusing because the output of 'True' should come after the TryParse and before the '$number'. Showing the output like this makes me think that $number contains both values.

Reply:
fixed